conversions for units

jread-usgs opened the pull request.

jread-usgs commented:
I screwed up pushing the remote branch, which is why I closed a bunch of PRs. didn't realize I did `git push -u upstream add_vars` instead of `git push -u origin add_vars` when I started this branch.

jread-usgs commented:
A couple notes here:

1. this does support multi-var conversions (e.g., `convert_sim_var(nc_file, crazy_var = temp-u_mean*1000)`)
2. I have it error when variables in the expression aren't in the simulation nc file. This is ok for now, but we should probably support a function name (such as `convert_sim_var(nc_file, temp.2 = convert(temp))`, where `convert` is a function in an accessible environment). 
3. This should have a standard evaluation supporting syntax, such as `add_sim_var(nc_file, var, unit, longname)` that takes a matrix for `var`, similar to that from `get_raw()`. This could be called on the inside of `convert_sim_var()`, which does this after capturing and dealing with the conversion function. 
4. I haven't tested this on any AED vars, but it should work. 
5. Not sure how things work out when you use different variables that are different dimensions, such as `wind` and `temp`. Not sure of the use case that would put 0D and 1D together, but maybe it'll happen?

jread-usgs commented:
Thoughts @lawinslow @hdugan?

hdugan commented:
I think this is great! Really good idea to add it to the ncdf file. Seems to work just fine with AED variables.

lawinslow commented on DESCRIPTION:
```diff
@@ -1,7 +1,7 @@
 Type: Package
 Title: glmtools
-Version: 0.13.2
+Version: 0.13.3
```
I'd say this should be a minor version change, not a patch version.

jread-usgs replied:
Agreed.

lawinslow commented on R/convert_sim_var.R:
```diff
@@ -0,0 +1,70 @@
+#'
+#' }
+#' 
+convert_sim_var <- function(nc_file='output.nc', ..., unit='', longname=''){
```
Will this support function calls as well? Like `tempF = c2f(temp)`. I can imagine a lot of unit conversions will use a package function.

jread-usgs replied:
See comments on the PR. It doesn't now, but it could with a minor modification. Basically, would need to remove the `stop` where it checks the parsed variables for being in the netcdf and errors if they aren't. 

Instead it would only pull the vars in netcdf into the `data` object, and leave the other names alone. Then the function would be called normally.

jread-usgs replied:
Might want to use `overwrite=FALSE` as an arg, so that you could overwrite an existing variable *if* you really wanted to.

jread-usgs commented on R/convert_sim_var.R:
```diff
@@ -0,0 +1,70 @@
+  fun.string <- deparse(convert[[1]]$expr)
+  variables <- strsplit(fun.string,"[^a-zA-Z_]")[[1]]
+  variables <- variables[variables != '']
+  vars.not.in <- variables[!variables %in% sim.vars]
```
to support the function use that Luke mentioned above, would want to change this to `nc.vars`, and do
```r
data <- lapply(nc.vars, function(v) get_raw(nc_file, v))
```
Then let the function call fail if bad names are used.

hdugan commented:
Like the idea of overwriting an existing variable. Especially if you didn't add units or longname the first time.

jread-usgs commented:
ok @lawinslow and @hdugan , I have addressed your comments in this PR - 

1. upped the minor version to 0.14.0
2. added supports for functions in the conversion and added a test for that.
3. added `overwrite` argument to the function, so you can overwrite a variable that is already in the file. and a test for that.

lawinslow commented:
Awesome.
